docs/painless/painless-guide/painless-datetime.asciidoc

```diff
@@ -0,0 +1,295 @@
+objective(s).
+
+==== Datetime API
+
+Datetimes in Painless use the standard Java libraries and are available through
+the Painless <<painless-api-reference-shared, Shared API>>. Most of the classes
+from the following Java packages are available to use in Painless scripts:
+
+* <<painless-api-reference-shared-java-time, java.time>>
+* <<painless-api-reference-shared-java-time-chrono, java.time.chrono>>
+* <<painless-api-reference-shared-java-time-format, java.time.format>>
+* <<painless-api-reference-shared-java-time-temporal, java.time.temporal>>
+* <<painless-api-reference-shared-java-time-zone, java.time.zone>>
+
+==== Datetime Parsing and Formatting
+
+Switching between different representations of datetimes is often necessary to
+achieve a script's objective(s). Datetime parsing is a switch from a long
+or String to a ZonedDateTime, and datetime formatting is a switch from a
+ZonedDateTime to a long or a String.
+
+A <<painless-api-reference-shared-DateTimeFormatter, DateTimeFormatter>> is a
+<<reference-types, reference type>> (object) that defines the allowed sequence
+of characters for a String representation of a datetime. Switching of datetime
+representations between a String and a ZonedDateTime and vice versa often
+requires a DateTimeFormatter. For more information about how to use a
+DateTimeFormatter see the
+{java11-javadoc}/java.base/java/time/format/DateTimeFormatter.html[Java documentation].
+
+===== Datetime Parsing Examples
+
+* parse from a long (milliseconds)
++
+[source,Painless]
+----
+long milliSinceEpoch = 434931330000L;
+Instant instant = Instant.ofEpochMilli(milliSinceEpoch);
+ZonedDateTime zdt = ZonedDateTime.ofInstant(instant, ZoneId.of('Z'));
+----
++
+* parse from a String (ISO 1806)
++
+[source,Painless]
+----
+String datetime = '1983-10-13T22:15:30Z';
+ZonedDateTime zdt = ZonedDateTime.parse(datetime);
+----
+Note the parse method uses ISO 1806 by default.
++
+* parse from a String (RFC 1123)
++
+[source,Painless]
+----
+String datetime = 'Thu, 13 Oct 1983 22:15:30 GMT';
+ZonedDateTime zdt = ZonedDateTime.parse(datetime,
+        DateTimeFormatter.RFC_1123_DATE_TIME);
+----
+Note the specification of a built-in DateTimeFormatter.
++
+* parse from a String (custom)
++
+[source,Painless]
+----
+String datetime = 'custom y 1983 m 10 d 13 22:15:30 Z';
+DateTimeFormatter dtf = DateTimeFormatter.ofPattern(
+        "'custom' 'y' yyyy 'm' MM 'd' dd HH:mm:ss VV");
+ZonedDateTime zdt = ZonedDateTime.parse(datetime, dtf);
+----
+Note the specification of a custom DateTimeFormatter.
++
+* parse from individual pieces
++
+[source,Painless]
+----
+int year = 1983;
+int month = 10;
+int day = 13;
+int hour = 22;
+int minutes = 15;
+int seconds = 30;
+int nanos = 0;
+String tz = 'Z';
+ZonedDateTime zdt = ZonedDateTime.of(
+        year, month, day, hour, minutes, seconds, nanos, ZoneId.of(tz));
+----
+
+===== Datetime Formatting Examples
+
+* format to a long (milliseconds)
```

Comment:
formatting usually means converting to a string. i think all these formatting examples can omit the "to a string" part. maybe the long to/from should be something separate, like conversion to/from epoch millis. Likewise, all the parsing examples could have the "from a string" removed

Reply:
Agreed.  Will spend some time fixing this.

Reply:
I've re-worked this, and hopefully, it's cleaner.